luminoth/tools/dataset/readers/object_detection/imagenet.py

```diff
@@ -0,0 +1,209 @@
+            else:
+                self._total = total_records
+
+        return self._total
+
+    @property
+    def classes(self):
+        if self._classes is None:
+            self._classes = sorted(list(self._wnids.values()))
+
+            # Choose random classes when limiting them
+            if self._limit_classes is not None and self._limit_classes > 0:
+                total_classes = min(len(self._classes), self._limit_classes)
+                self._classes = sorted(
+                    random.sample(self._classes, total_classes)
+                )
+
+        return self._classes
+
+    def iterate(self):
+        for image_id in self._get_record_names():
+            # Ignore image_id is using only_filename filter and it's different.
+            if self._only_filename and image_id != self._only_filename:
+                continue
+
+            if self.yielded_records == self.total:
+                # Finish iteration based on predefined total.
+                return
+
+            try:
+                annotation_path = self._get_image_annotation(image_id)
+                image_path = self._get_image_path(image_id)
+                # Read both the image and the annotation into memory.
+                annotation = read_xml(annotation_path)
+                image = read_image(image_path)
+            except tf.errors.NotFoundError:
+                tf.logging.debug(
+                    'Error reading image or annotation for "{}".'.format(
+                        image_id))
+                self.errors += 1
+                continue
+
+            objects = annotation.get('object')
+            if objects is None:
+                # If there's no bounding boxes, we don't want it.
+                continue
+
+            # TODO: consider alternatives to using Pillow here.
+            image_pil = Image.open(image_path)
+            width = image_pil.width
+            height = image_pil.height
+            image_pil.close()
+
+            gt_boxes = []
+            for b in annotation['object']:
+                try:
+                    label_id = self.classes.index(self._wnids[b['name']])
+                except ValueError:
+                    continue
+
+                (xmin, ymin, xmax, ymax) = self._adjust_bbox(
+                    xmin=int(b['bndbox']['xmin']),
+                    ymin=int(b['bndbox']['ymin']),
+                    xmax=int(b['bndbox']['xmax']),
+                    ymax=int(b['bndbox']['ymax']),
+                    old_width=int(annotation['size']['width']),
+                    old_height=int(annotation['size']['height']),
+                    new_width=width, new_height=height
+                )
+
+                gt_boxes.append({
+                    'label': label_id,
+                    'xmin': xmin,
+                    'ymin': ymin,
+                    'xmax': xmax,
+                    'ymax': ymax,
+                })
+
+            if len(gt_boxes) == 0:
+                continue
```

Comment:
Does this mean we don't care about examples without a GT?
Is this always the case? Should it depend on the algorithm used?

Faster R-CNN wouldn't work because we need positive examples, but some algorithms may work with minibatches of > 1 images so it would make sense to constrain this while doing dataset ingestion here.

If this makes sense, this could be a property of the reader (initialized to skip samples with no GT, or not).

Reply:
When filtering classes we can end up with empty images. We ignore those images.